add TableDiff.getSummary() for summary stats of diff

paulfitz opened the pull request:
After `TableDiff.hilite` has been used to produce a diff, the `getSummary` method can be called to get an overall summary of what changed, including counts of row insertions, deletions, etc.

Motivated by https://github.com/paulfitz/daff/issues/87.  Does this look like it would meet your need @gwarnes-mdsol?

gwarnes-mdsol commented:
Looks great @paulfitz!  I'll take it for a spin.

gwarnes-mdsol commented:
Would it be possible to add 'col_updates' containing the number of columns containing changes?

paulfitz commented:
Yes, that should be doable.  Unlike all the other categories, it isn't
something that TableDiff directly knows about, but there would be no harm
adding it.  So any column which contains at least one modified cell would
be considered "changed"?  Row insertions and deletions would be ignored I
guess?  Since otherwise all columns would be changed if any row
insertion/deletion were present, which wouldn't be very informative.

On Mon, Mar 27, 2017 at 12:15 PM, Gregory R. Warnes <
notifications@github.com> wrote:

> Would it be possible to add 'col_updates' containing the number of columns
> containing changes?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulfitz/daff/pull/88#issuecomment-289503882>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHOXz-RVbCWyykkB_rWfMsXkl2-TPxjks5rp-C1gaJpZM4MpGWC>
> .
>

gwarnes-mdsol commented:
You got it.  Changed columns are columns containing a 'modified' cell (i.e.
marked with '->').  ON this end I've been  grep'ing through each column for
the presence of the change marker.



On Mon, Mar 27, 2017 at 2:16 PM, Paul Fitzpatrick <notifications@github.com>
wrote:

> Yes, that should be doable. Unlike all the other categories, it isn't
> something that TableDiff directly knows about, but there would be no harm
> adding it. So any column which contains at least one modified cell would
> be considered "changed"? Row insertions and deletions would be ignored I
> guess? Since otherwise all columns would be changed if any row
> insertion/deletion were present, which wouldn't be very informative.
>
> On Mon, Mar 27, 2017 at 12:15 PM, Gregory R. Warnes <
> notifications@github.com> wrote:
>
> > Would it be possible to add 'col_updates' containing the number of
> columns
> > containing changes?
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/paulfitz/daff/pull/88#issuecomment-289503882>
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_paulfitz_daff_pull_88-23issuecomment-2D289503882-253E&d=DwQFaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=KKpQIICbmJRQqALr_78PgbpYVY1LbNAc_S50a49GBq8&s=nFyw-ebr5GEN-LjWPpixsxAFFXlgSuqBldhq3LNIfb0&e=>,
> or mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/AAHOXz-RVbCWyykkB_
> rWfMsXkl2-TPxjks5rp-C1gaJpZM4MpGWC>
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAHOXz-2DRVbCWyykkB-5FrWfMsXkl2-2DTPxjks5rp-2DC1gaJpZM4MpGWC-253E&d=DwQFaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=KKpQIICbmJRQqALr_78PgbpYVY1LbNAc_S50a49GBq8&s=_W9j1mZkJKTi6z19FnJ4bTva3V28yyojagRrJOMJOEI&e=>
> > .
>
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_paulfitz_daff_pull_88-23issuecomment-2D289538672&d=DwMFaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=KKpQIICbmJRQqALr_78PgbpYVY1LbNAc_S50a49GBq8&s=nuXHelWECeAWyE_vw--Jo2LyzNNNnu-I-RMPUIH9a00&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AVNo-2DnJSuV8XyoTxzSixMeRIl2T6hOzxks5rp-5F0PgaJpZM4MpGWC&d=DwMFaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=KKpQIICbmJRQqALr_78PgbpYVY1LbNAc_S50a49GBq8&s=A5FHR173VArNYWM3w5csQG-VyrVhbDDTxQ_XE2QMODA&e=>
> .
>



-- 
*Gregory R. Warnes, Ph.D.* | Principal Data Scientist | Medidata Solutions
<http://www.mdsol.com>
350 Hudson Street, 9th Floor, New York, NY 10014
gwarnes@mdsol.com | direct: +1 646 362 2829 | main: +1 212 918 1800

Follow Us: Twitter <http://twitter.com/#!/Medidata> | LinkedIn
<http://www.linkedin.com/company/medidata-solutions> | Read Our Blog: Geeks
Talk Clinical <http://geekstalkclinical.mdsol.com>

The information in this transmittal and any attachments are confidential
and intended only for the recipient(s) listed above.
You are hereby notified that any unauthorized distribution or copying of
this transmittal or its attachments is prohibited.
If you have received this transmittal in error, please notify the sender.

gwarnes-mdsol commented:
This is working great for me, thanks!

At the moment, I'm omitting the # of columns containing changes from my summary, so I'll be looking forward to that enhancement. :-)

paulfitz commented:
Thanks for the feedback, glad it is working!  Adding # of changed columns
is on my todo list, should get to it within a week.

On Thu, Mar 30, 2017 at 9:34 AM, Gregory R. Warnes <notifications@github.com
> wrote:

> This is working great for me, thanks!
>
> At the moment, I'm omitting the # of columns containing changes from my
> summary, so I'll be looking forward to that enhancement. :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/paulfitz/daff/pull/88#issuecomment-290412048>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHOXyycE8VpVz4uNQlK-nJOVcuqpSCdks5rq69lgaJpZM4MpGWC>
> .
>

paulfitz commented:
@gwarnes-mdsol there's now a `col_updates` summary field, let me know if it does what you need.

gwarnes-mdsol commented:
Perfect, thanks!

gwarnes-mdsol commented:
One bug:  When columns are reordered *and* added, these summaries are (usually) incorrect.  
IE,
```
> iris2 <- cbind(iris[, c(1:3,5,4)], new=1:nrow(iris))
> summary(diff_data(iris, iris2))

Data diff:
 Comparison: ‘iris’ vs. ‘iris2’ 
        #       Modified Reordered Deleted Added
Rows    150     0        0         0       0    
Columns 5 --> 6 0        1         0       0    
# The reordered count is (in this case correct), but the added  count is not.
```

In some other cases the added columns are included in the 'modified' count.

paulfitz commented:
@gwarnes-mdsol I added tests for reorders+inserts and fixed a bug that could miss out inserts.  Thanks for reporting!  For the `added columns are improperly included in the 'reordered' count` case, yes a simple example would be great, just for clarity.

gwarnes-mdsol commented:
The current code works properly:

```
> iris2 <- cbind(iris[, c(1:3,5,4)], new=1:nrow(iris))
> summary(diff_data(iris, iris2))

Data diff: ‘iris’ vs. ‘iris2’ 
        #       Modified Reordered Deleted Added
Rows    150     0        0         0       0    
Columns 5 --> 6 0        2         0       1    
```

gwarnes-mdsol commented:
@paulfitz Is there any reason not to merge this into master?

paulfitz commented:
@gwarnes-mdsol no, no reason, thanks for nudge!

gwarnes-mdsol commented:
:-)


On Fri, Jun 23, 2017 at 5:59 PM Paul Fitzpatrick <notifications@github.com>
wrote:

> @gwarnes-mdsol
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_gwarnes-2Dmdsol&d=DwMCaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=xthw4CTM0fvR2az2FwEac3c7UFXPWWotFTDwem0-tco&s=dbWPJprzGSKXFLonouAKLX7VqA8B-T5XGhHCtchzG9U&e=>
> no, no reason, thanks for nudge!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_paulfitz_daff_pull_88-23issuecomment-2D310782321&d=DwMCaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=xthw4CTM0fvR2az2FwEac3c7UFXPWWotFTDwem0-tco&s=yv57Jex41JwXFVo6CJEkf5yq7KTN9yzrBQ-cFUbct3o&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AVNo-2DnY5eDyA3OBhP1EmvOEPPcx5o1eOks5sHDVFgaJpZM4MpGWC&d=DwMCaQ&c=fi2D4-9xMzmjyjREwHYlAw&r=PUo6rYTmGeWkBJepZc1UHw629PctwMWQF8I3RzhQlL8&m=xthw4CTM0fvR2az2FwEac3c7UFXPWWotFTDwem0-tco&s=KR3hvLimsQYyhJ8GtJENGhRTlQdLTLpjnfbCuxPCXrs&e=>
> .
>
-- 
*Gregory R. Warnes, Ph.D.* | Principal Data Scientist | Medidata Solutions
<http://www.mdsol.com>
350 Hudson Street, 9th Floor, New York, NY 10014
gwarnes@mdsol.com | direct: +1 646 362 2829 | main: +1 212 918 1800

Follow Us: Twitter <http://twitter.com/#!/Medidata> | LinkedIn
<http://www.linkedin.com/company/medidata-solutions> | Read Our Blog: Geeks
Talk Clinical <http://geekstalkclinical.mdsol.com>

The information in this transmittal and any attachments are confidential
and intended only for the recipient(s) listed above.
You are hereby notified that any unauthorized distribution or copying of
this transmittal or its attachments is prohibited.
If you have received this transmittal in error, please notify the sender.
